### PR TITLE
[Discuss] Async use cases

### PR DESCRIPTION
**Pull request by bendk**
I don't know if this document should be added to the docs section, the main reason I'm opening the PR is so we can discuss the use cases more.

### CODE REVIEW

File: docs/async-use-cases.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+This is the same as the last one, except the call is async.  The main use case is probably the same
+as the previous section, except when Rust is using an event loop rather than a thread pool.
+
+# New features
+
+Based on the above analysis, I believe there are 2 features that we should consider adding:
+
+  - Rust code making an async call into a CallbackInterface / Foreign code
+    making an async into Rust. I think this could be implemented by having
+    UniFFI generate something like the hand-written demo from #1252.
```
Comment (bendk) [inline docs/async-use-cases.md:90]: This seems like a nice feature, but I'm not sure it's necessary.  I think we could implement the Nimbus work with hand-written code.  Maybe we could just start there and see if there's more demand before generating that code with UniFFI.

File: docs/async-use-cases.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  - Handling CallbackInterface calls from a Rust-based thread. I'm not exactly sure how this would work, there's at least 2 options here:
+    - Handle everything on the foreign side. In the generate code we currently register a callback to invoke a CallbackInterface call. We might be able to update that code so that it schedules the call to run on the correct thread. But this assumes that it's safe to call that callback on the Rust thread. Is that true for all of our current languages? Are we okay with adding this requirement for future languages?
+    - Use a queue plus a waker. Push the CallbackInterface call to a queue, signal the foreign side (maybe writing a byte to a pipe or socket), then the foreign side would wake up and try to read from the queue. This seems more complicated than the first system, but might let us support more foreign languages.
```
Comment (bendk) [inline docs/async-use-cases.md:93]: This feature seems necessary if we ever want to make a call from a Rust controlled (or C++ controlled) thread into JS.  Maybe we want that for the Nimubs dekstop work, but I'm not sure.

File: docs/async-use-cases.md (hunk 1 above)
Comment (badboy) [reply docs/async-use-cases.md:90]: I agree that hand-writing & using it that way seems sensible. That's how UniFFI was born to begin with.

**Comment by bendk** [comment]
I think we have a good plan here, no need to keep this PR open longer.

**Pull request closed without merging**